[script.service.ltvpl] 1.0.0

jmooremcc opened the pull request:
### Description
<!--- Provide a short summary of submitted add-on in case it's a new addition. -->
The Live TV Playlist add-on allows PVR users to create a list of their favorite live TV shows. It will automatically change the channel to your favorite show a few seconds before your program is scheduled to start. Among its many features is the ability of the playlist to easily accommodate recurring TV shows, such as those that air on a daily or weekly basis as well as on weekdays and weekends. The Live TV Playlist add-on makes watching TV a more enjoyable experience.
NOTE: A PVR add-on is required to watch live TV.
<!--- If it's plugin update only highlight biggest changes if needed. -->
<!--- Make sure you follow the checklist below before finalising your pull-request. -->
### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [X] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

jmooremcc commented:
@Rechi, @enen92  Any idea why it's taking so long for my add-on to be approved for the repo? It's been a few months now and I haven't received any feedback indicating what if any are the problems holding up my add-on's inclusion into the official repository.

ronie commented:
you require script.module.enum34 but it's not available in our repo.

sharing both resources and resources/lib as python modules is not a thing you should do.
python modules are only used to share functionalities with other addons.

jmooremcc commented:
> you require script.module.enum34 but it's not available in our repo.
> 
> sharing both resources and resources/lib as python modules is not a thing you should do.
> python modules are only used to share functionalities with other addons.

You're correct which is why I previously submitted PR #1108 [script.module.enum34]]. They've been dragging their feet on accepting that PR which prevents this PR (1172) from having this conflict resolved. This has been going on for months now. Initially I had the enum module embedded as part of my add-on and was advised by @Lunatixz to put the enum34 module, which was created by the enum author, as a separate module. At this point I don't know what to do because I keep getting conflicting advice.
My original design, as I said previously, included the enum module as part of the add-on. My intent was to remove that module when running under Python 3.x since it won't be needed. I've tested my beta add-on with users in Europe and I know it works well. This add-on has been under development for 2 years and I'm very confident about how well it works. In fact, a number of users have told me that they thought it was a very useful add-on.

Anything you can do to help me get this in the Kodi repository will be greatly appreciated.

jmooremcc commented:
@ronie Can you tell me what pending check Travis Buddy is waiting on.
I've removed the requirement for the script.module.enum34

enen92 commented:
Closing as it looks stalled. Please resubmit if the interest persists.

jmooremcc commented:
Hi,

Can you tell me why it’s stalled?
It passed all the tests as far as I can see and no one has communicated to me what if any problems exist.

Thank you.

John

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________
From: enen92 <notifications@github.com>
Sent: Sunday, February 9, 2020 11:31:37 AM
To: xbmc/repo-scripts <repo-scripts@noreply.github.com>
Cc: jmooremcc <jmooremcc@hotmail.com>; Author <author@noreply.github.com>
Subject: Re: [xbmc/repo-scripts] [script.service.ltvpl] 1.0.0 (#1172)


Closing as it looks stalled. Please resubmit if the interest persists.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/xbmc/repo-scripts/pull/1172?email_source=notifications&email_token=ADESKBL6P4KNCPL6RBVAEPTRCAVWTA5CNFSM4I4P5U7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELGRG4A#issuecomment-583865200>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADESKBO57DTVK2EAXNQJ4MLRCAVWTANCNFSM4I4P5U7A>.

enen92 commented:
Sorry looks like no one has reviewed the pull request. I'll do it myself.

enen92 commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,57 @@
+<addon id="script.service.ltvpl" name="Live TV Playlist" provider-name="jmooremcc" version="1.0.0">
+  <requires>
+    <import addon="xbmc.python" version="2.26.0" />
+    <import addon="xbmc.addon" version="17.0.0" />
```
`xbmc.addon` dependency not needed, please remove

jmooremcc replied:
Removed

enen92 commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,57 @@
+  <requires>
+    <import addon="xbmc.python" version="2.26.0" />
+    <import addon="xbmc.addon" version="17.0.0" />
+    <import addon="xbmc.gui" version="5.14.0" />
```
`xbmc.gui` dependency not needed, please remove

jmooremcc replied:
Removed

enen92 commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,57 @@
+    <import addon="xbmc.gui" version="5.14.0" />
+  </requires>
+  <extension library="service.py" point="xbmc.service" />
+  <extension library="resources/lib" point="xbmc.python.module" />
```
something's wrong here, you're defining 2 modules. Maybe the resources one is not needed. Can you clarify?

jmooremcc replied:
Removed resources/lib

enen92 commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,57 @@
+    <import addon="xbmc.addon" version="17.0.0" />
+    <import addon="xbmc.gui" version="5.14.0" />
+  </requires>
+  <extension library="service.py" point="xbmc.service" />
```
this is the main addon type. Maybe you should consider moving the script extension point to the top (because it will  be listed under programs in the addon browser)

jmooremcc replied:
Moved to top

enen92 commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,57 @@
+      Der Manager für TV-Wiedergabelisten stellt einen Service für Clients zur Verfügung, um Fernsehsendungen zu planen. Eine Wiedergabeliste für TV-Programme ist eine Liste von Kanalwechselbefehlen, die zu einer bestimmten Zeit stattfinden. Ein Element der Wiedergabeliste besteht aus einer Alarmzeit, einem Kanal und einer Wiederholungsoption. Der Manager verwaltet eine Datenbank mit Wiedergabeli ...
+    </description>
+    <platform>all</platform>
+    <license>GNU GENERAL PUBLIC LICENSE. Version 3, June 2007</license>
```
Please use the SPDX identifier equivalent GPL-3.0-only or GPL-3.0-or-later depending on the license (see https://spdx.org/licenses/)

jmooremcc replied:
Done

enen92 commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,57 @@
+    <assets>
+      <icon>icon.png</icon>
+      <fanart>fanart.jpg</fanart>
+      <screenshot>resources/screenshots/screenshot000.png</screenshot>
```
please use only a maximum of 4 screenshots (we'll be shipping this from our infrastructure)

jmooremcc replied:
Done.
Also updated assets section to reflect only 4 screenshots.

enen92 commented on script.service.ltvpl/keymapper.py:
```diff
@@ -0,0 +1,175 @@
+
+__Version__ = "1.0.1"
+
+keymapfile = os.path.join(KEYMAPS_USERDATA_FOLDER,"gen.xml")
```
sorry this is not allowed in the repository. You can't edit such xml files directly from python. The only alternative is to provide instructions on your addon on how to use keymap editor addon.

jmooremcc replied:
My add-on requires the [**script.module.enum34]** (PR 1108) module in order to run. The enum module is not included in Python 2 but is included in Python 3. This PR was closed recently because no add-on in the repository references it. 

It is my intention that this version of my add-on will only be qualified to run under Leia. I will be removing the requirement for the [**script.module.enum34]** module for Matrix since it will be running Python 3.

jmooremcc replied:
Removed

enen92 commented on script.service.ltvpl/resources/PL_Server.py:
```diff
@@ -0,0 +1,477 @@
+            seconds = tomorrow - datetime.now()
+            self.dailyMaintThread = Timer(seconds.total_seconds(), self.dataSet.verifyDataset)
+            self.dailyMaintThread.start()
+            time.sleep(1.0)
```
use xbmc.sleep or xbmc.Monitor().waitForAbort

jmooremcc replied:
This code also runs in a non-Kodi environment which is why time.sleep() is used.

enen92 commented on script.service.ltvpl/resources/colours/Color.xml:
```diff
@@ -0,0 +1,144 @@
+<colors>
```
I didn't know this was supported in scripts. @ronie ?

ronie replied:
nope, it isn't.

jmooremcc replied:
Removed

enen92 commented on script.service.ltvpl/resources/data/dataFileLocation.py:
```diff
@@ -0,0 +1,2 @@
+#
+dataFilePath = r"D:\My Documents\GitHub\LiveTV_PlayList\script.service.ltvpl\resources\data\LTVPL.pkl"
```
hardcoded path, this won't work

jmooremcc replied:
This absolute path is replaced at runtime by a relative path when the code runs under the Kodi environment. This code also runs under Windows which is why you see that absolute path. It does work on all my systems running Kodi ( Raspberry Pi & Windows).

enen92 commented on script.service.ltvpl/resources/data/dataFileLocation2.py:
```diff
@@ -0,0 +1 @@
+dataFilePath = r"../resources\data/service.ltvpl\LTVPL.pkl"
```
same

enen92 commented on script.service.ltvpl/resources/data/debugFileLocation.py:
```diff
@@ -0,0 +1 @@
+DEBUGCACHEFILE = r"D:\My Documents\GitHub\LiveTV_PlayList\script.service.ltvpl\resources\data\debugcache.json"
```
same

jmooremcc replied:
This absolute path is replaced at runtime by a relative path when the code runs under the Kodi environment.

enen92 commented on script.service.ltvpl/resources/lib/Data/PL_DataSet.py:
```diff
@@ -0,0 +1,684 @@
+        DbgPrint("***LastChannel: {}".format(self.lastChannel))
+        if KODI_ENV and self.lastChannel is None:
+            DbgPrint("*****LastChannel->: {} isNone:{}".format(self.lastChannel, self.lastChannel is None))
+            xbmcgui.Dialog().notification(LTVPL, "Scanning Channel List...")
```
string needs to be localized as it is presented to the user

jmooremcc replied:
done!

enen92 commented on script.service.ltvpl/resources/lib/Data/fileManager.py:
```diff
@@ -0,0 +1,156 @@
+            os.remove(f)
+
+    def addSentinel(self, dstfile):
+        fp = open(self.sentinelFile, WRITEMODE)
```
use python `with` statement

jmooremcc replied:
done

enen92 commented on script.service.ltvpl/resources/lib/Data/fileManager.py:
```diff
@@ -0,0 +1,156 @@
+                count -= 1
+
+    def restoreFromSentinel(self):
+        fp = open(self.sentinelFile, 'r')
```
same

jmooremcc replied:
done

enen92 commented on script.service.ltvpl/resources/lib/KodiLib/kodiflags.py:
```diff
@@ -0,0 +1,11 @@
+
+try:
```
a simple try/importError is enough

jmooremcc replied:
done

enen92 commented on script.service.ltvpl/resources/lib/Utilities/VirtualEvents.py:
```diff
@@ -0,0 +1,47 @@
+#VirtualEvents.py
+
+try:
+    from Tkinter import *
```
I hope this is only for compatibility for environments outside of kodi

jmooremcc replied:
You're correct with your assumption. This back end code also runs under Windows.

enen92 commented on script.service.ltvpl/resources/lib2/enum/LICENSE:
```diff
@@ -0,0 +1,32 @@
+Copyright (c) 2013, Ethan Furman.
```
this has to be submitted as an external python script.module

jmooremcc replied:
My add-on requires the [script.module.enum34] (PR 1108) module in order to run. The enum module is not included in Python 2 but is included in Python 3. This PR was closed recently because no add-on in the repository references it.

It is my intention that this version of my add-on will only be qualified to run under Leia. I will be removing the requirement for the [script.module.enum34] module for Matrix since it will be running Python 3.

jmooremcc replied:
Removed from my add-on.

enen92 commented on script.service.ltvpl/util.py:
```diff
@@ -0,0 +1,140 @@
+
+HOME    =  ADDON_PATH
+ROOT    =  ADDON.getSetting('FOLDER')
+PROFILE =  os.path.join(ROOT, 'Super Favourites')
```
what is this?

jmooremcc replied:
Fixed.

ronie commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,57 @@
+  </requires>
+  <extension library="service.py" point="xbmc.service" />
+  <extension library="resources/lib" point="xbmc.python.module" />
+  <extension library="resources" point="xbmc.python.module" />
```
neither should be needed. this exposes the entire resource dir as a python library to the outside world.
the xbmc.python.module extension point is intended for script.module.* addons.

the only exception would be if you need to share a part of your code with another addon, but i don't think that's the case here.

jmooremcc replied:
When I remove `<extension library="resources" point="xbmc.python.module" />` from my addon.xml file, none of my textures  appear when the gui window is displayed. 
The only way I know how to make it work is by including that line in my addon.xml file.
How does Kodi know where to search for my textures?
Is there another way for me to tell Kodi where my textures are?

ronie commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,57 @@
+      </item>
+    </menu>
+  </extension>
+  <extension debugging="false" effectslowdown="0.75" point="xbmc.gui.skin">
```
this is only for skins. you're not a skin.

jmooremcc replied:
removed

ronie commented on script.service.ltvpl/resources/skins/Default/skin.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
this file is not needed. skin.xml was deprecated many years ago.

jmooremcc replied:
removed

ronie commented on script.service.ltvpl/resources/skins/Default/720p/Includes.xml:
```diff
@@ -0,0 +1,31 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
addons can not use Includes files

jmooremcc replied:
removed

ronie commented on script.service.ltvpl/resources/skins/Default/720p/BusyDialog.xml:
```diff
@@ -0,0 +1,64 @@
+	<animation effect="fade" time="100">WindowOpen</animation>
+	<animation effect="fade" time="300">WindowClose</animation>
+	<coordinates>
+		<system>1</system>
```
this is not used anymore by kodi

jmooremcc replied:
removed

ronie commented on script.service.ltvpl/resources/skins/Default/720p/BusyDialog2.xml:
```diff
@@ -0,0 +1,54 @@
+				<animation effect="fade" start="100" end="0" time="240" condition="Window.IsVisible(fullscreenvideo)">Conditional</animation>
+			</control>
+			<control type="group">
+				<depth>DepthMax</depth>
```
there is no guarantee `DepthMax` is defined in the skin you're using

jmooremcc replied:
removed

ronie commented on script.service.ltvpl/resources/skins/Default/720p/captureEpgItem.xml:
```diff
@@ -0,0 +1,338 @@
+    <posx>300</posx>
+    <posy>200</posy>
+    <width>350</width>
+    <height>200</height>
```
width / height are not supported by `coordinates`

jmooremcc replied:
removed

ronie commented on script.service.ltvpl/resources/skins/Default/720p/contextmenu.xml:
```diff
@@ -0,0 +1,146 @@
+          <posy>0</posy>
+          <width>250</width>
+          <height>38</height>
+          <font>fontContextMenu</font>
```
there's no guaranty `fontContextMenu` is defined by the skin you're using

jmooremcc replied:
removed

ronie commented on script.service.ltvpl/resources/skins/Default/720p/custom_help.xml:
```diff
@@ -0,0 +1,175 @@
+      <top>0</top>
+      <width>900</width>
+      <height>500</height>
+      <aspectratio>1</aspectratio>
```
1 is not a valid value for aspectratio

jmooremcc replied:
fixed

ronie commented:
some general comments regarding the skin part in your addon:
- prefix all xml files with the name of your addon, to avoid clashes with xml files shipped by a skin.
eg. ltvpl-captureEpgItem.xml
- prefix all images with the name of your addon, to avoid clashes with images shipped by a skin.
eg. ltvpl-button-focus.png
- do not use includes  (`<include>foo</include>`) there is no guarantee such an include is defined by the skin you're using
- don't include unneeded/unused files
eg. DialogBack.png.old / radialmask.psd


two things i've noticed after installing the addon:
- it does not show up in the addon manager (neither in programs, nor in services)
- kodi locks up on exit due to your service crashing on shutdown
```
2020-02-10 00:12:05.328 T:11314   DEBUG: CPythonInvoker(1, /home/ronie/.kodi/addons/script.service.ltvpl/service.py): trigger Monitor abort request
2020-02-10 00:12:05.328 T:11367   DEBUG: vacationmode1:false
2020-02-10 00:12:05.431 T:11367   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <class 'AttributeError'>
                                            Error Contents: module 'time' has no attribute 'clock'
                                            Traceback (most recent call last):
                                              File "/home/ronie/.kodi/addons/script.service.ltvpl/service.py", line 237, in <module>
                                                server.stopServer()
                                              File "/home/ronie/.kodi/addons/script.service.ltvpl/resources/PL_Server.py", line 129, in stopServer
                                                startTimer()  # measure time to stop server
                                              File "/home/ronie/.kodi/addons/script.service.ltvpl/resources/lib/Utilities/DebugPrint.py", line 211, in startTimer
                                                _startTime = time.clock()
                                            AttributeError: module 'time' has no attribute 'clock'
                                            -->End of Python script error report<--
```

jmooremcc commented:
Thank you ronie for this valuable feedback.
I will get to work on resolving these issues this week.

John

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

From: ronie<mailto:notifications@github.com>
Sent: Sunday, February 9, 2020 7:19 PM
To: xbmc/repo-scripts<mailto:repo-scripts@noreply.github.com>
Cc: jmooremcc<mailto:jmooremcc@hotmail.com>; Author<mailto:author@noreply.github.com>
Subject: Re: [xbmc/repo-scripts] [script.service.ltvpl] 1.0.0 (#1172)


some general comments regarding the skin part in your addon:

  *   prefix all xml files with the name of your addon, to avoid clashes with xml files shipped by a skin.
eg. ltvpl-captureEpgItem.xml
  *   prefix all images with the name of your addon, to avoid clashes with images shipped by a skin.
eg. ltvpl-button-focus.png
  *   do not use includes (<include>foo</include>) there is no guarantee such an include is defined by the skin you're using
  *   don't include unneeded/unused files
eg. DialogBack.png.old / radialmask.psd

two things i've noticed after installing the addon:

  *   it does not show up in the addon manager (neither in programs, nor in services)
  *   kodi locks up on exit due to your service crashing on shutdown

2020-02-10 00:12:05.328 T:11314   DEBUG: CPythonInvoker(1, /home/ronie/.kodi/addons/script.service.ltvpl/service.py): trigger Monitor abort request

2020-02-10 00:12:05.328 T:11367   DEBUG: vacationmode1:false

2020-02-10 00:12:05.431 T:11367   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--

                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!

                                            Error Type: <class 'AttributeError'>

                                            Error Contents: module 'time' has no attribute 'clock'

                                            Traceback (most recent call last):

                                              File "/home/ronie/.kodi/addons/script.service.ltvpl/service.py", line 237, in <module>

                                                server.stopServer()

                                              File "/home/ronie/.kodi/addons/script.service.ltvpl/resources/PL_Server.py", line 129, in stopServer

                                                startTimer()  # measure time to stop server

                                              File "/home/ronie/.kodi/addons/script.service.ltvpl/resources/lib/Utilities/DebugPrint.py", line 211, in startTimer

                                                _startTime = time.clock()

                                            AttributeError: module 'time' has no attribute 'clock'

                                            -->End of Python script error report<--

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/xbmc/repo-scripts/pull/1172?email_source=notifications&email_token=ADESKBNOTSMSZCPLVIVCETLRCCMQ5A5CNFSM4I4P5U7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELG4YUA#issuecomment-583912528>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADESKBLF33PX2PZAMIJJVWTRCCMQ5ANCNFSM4I4P5U7A>.

BubbeleM commented:
Na, geil? Ihr dummen geile Nerds..
😂😂😂

Am 15.02.2020 um 21:46 schrieb jmooremcc <notifications@github.com>:

﻿

@jmooremcc commented on this pull request.

________________________________

In script.service.ltvpl/resources/skins/Default/720p/BusyDialog2.xml<https://github.com/xbmc/repo-scripts/pull/1172#discussion_r379855113>:

> +     <animation effect="fade" end="100" time="560">WindowOpen</animation>
+       <animation effect="fade" start="100" end="0" time="240">WindowClose</animation>
+       <zorder>2</zorder>
+       <controls>
+               <control type="group">
+                       <visible>String.IsEmpty(Window(Home).Property(script.cinemavision.running))</visible>
+                       <animation effect="fade" time="400">VisibleChange</animation>
+                       <visible>!Window.IsActive(startup)</visible>
+                       <control type="image">
+                               <texture>colors/black.png</texture>
+                               <include>FullScreenDimensions</include>
+                               <animation effect="fade" start="100" end="70" time="0" condition="true">Conditional</animation>
+                               <animation effect="fade" start="100" end="0" time="240" condition="Window.IsVisible(fullscreenvideo)">Conditional</animation>
+                       </control>
+                       <control type="group">
+                               <depth>DepthMax</depth>


removed

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/xbmc/repo-scripts/pull/1172?email_source=notifications&email_token=AN6KJEHHWA6DXMH6G3YJXTLRDBIDNA5CNFSM4I4P5U7KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVVXX3Q#discussion_r379855113>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AN6KJEC7FMEI5UP53YL5TUDRDBIDNANCNFSM4I4P5U7A>.

jmooremcc commented:
I've updated my add-on per your requested changes.

basrieter commented:
> I've updated my add-on per your requested changes.
> The check failure is due to the non-availability of the enum34 module.

Indeed, we don't seem have that one. There have been attempts (https://github.com/xbmc/repo-scripts/pull/1108) but not successful. I would try to remove the dependency and update your PR.

basrieter commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,46 @@
+  <extension library="service.py" point="xbmc.service" />
+  <requires>
+    <import addon="xbmc.python" version="2.26.0" />
+    <import addon="script.module.enum34" version="1.1.6" />
```
`script.module.enum34` is not available.

jmooremcc replied:
I submitted the module for enum34 and no one has reviewed it and told me why it has not been accepted into the repository. The enum module is not included in the Pyton 2.7 standard library but is included in the Python 3.x standard library. My add-on requires the enum module.

basrieter commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,46 @@
+    <email>jmooremcc@hotmail.com</email>
+    <assets>
+      <icon>icon.png</icon>
+      <fanart>fanart.jpg</fanart>
```
file missing

jmooremcc replied:
I will fix this problem in my next update.
The files were renamed per an earlier recommendation but I forgot to update the assets entry in addon.xml.

basrieter commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,46 @@
+    <source>https://github.com/jmooremcc/Live_TV_Playlist</source>
+    <email>jmooremcc@hotmail.com</email>
+    <assets>
+      <icon>icon.png</icon>
```
file missing

jmooremcc replied:
I've modified PR #1108 to be compatible with Python 2 & 3 per your suggestion.
I am updating this PR to utilize the updated enum module and fixed the 2 missing file references you pointed out in my addon.xml file.

jmooremcc commented:
> > I've updated my add-on per your requested changes.
> > The check failure is due to the non-availability of the enum34 module.
> 
> Indeed, we don't seem have that one. There have been attempts (#1108) but not successful. I would try to remove the dependency and update your PR.

I submitted a PR for a compatible enum34 module and no one has reviewed it and given me any feedback on it. My add-on requires the enum module when running under Python 2.7 because enum is not part of the standard library. Python 3.x does have enum as part of the standard library.

I am requesting that you review my PR 1108 for the enum34 module and tell why it has not been accepted. If there is another enum module available in the repository, please let me know so that I can use that instead.

basrieter commented:
So the bottom line for https://github.com/xbmc/repo-scripts/pull/1108 is that it is not compatible with both Python 2 and 3 in the way dagwieers explained; add-on developers that target python 3.4 and python 2 have no way of distinguishing between the Py3 built-in `enum` and `enum34` because you packaged it as `enum` in the `lib` folder. 

I would suggest you make it available as `enum34` as Dag suggested. That way it can be used as 

``` 
try:  # Python 3
    from enum import Enum
except ImportError:
    from enum34 import Enum
```

jmooremcc commented:
> The enum34 dependency is currently blocking your submission. I would just try to work around it or submit a proper Py2/Py3 compatible module add-on.
> 
> Als the icon and fanart are missing.
> 
> Also:
> 
> > WARN: Complex entry point. Check: service.py | Counted lines: 167 | Lines allowed: 15
> > WARN: Complex entry point. Check: addon.py | Counted lines: 646 | Lines allowed: 15

I've made the suggested changes to PR #1108 enum34 module to make it compatible with Python 2 and 3. Specifically, I renamed the enum directory to enum34.
Please reopen PR #1108 for acceptance into the repository.

jmooremcc commented:
I might have misunderstood a suggested change to icon and fanart. It was suggested that I prepend ltvpl- to all my image files so I renamed fanart to ltvpl-fanart and icon to ltvpl-icon. If this was not correct, please let me know.

basrieter commented:
> I might have misunderstood a suggested change to icon and fanart. It was suggested that I prepend ltvpl- to all my image files so I renamed fanart to ltvpl-fanart and icon to ltvpl-icon. If this was not correct, please let me know.

The thing is that the fanart and icon are actually missing from your submission!

jmooremcc commented:
> > I might have misunderstood a suggested change to icon and fanart. It was suggested that I prepend ltvpl- to all my image files so I renamed fanart to ltvpl-fanart and icon to ltvpl-icon. If this was not correct, please let me know.
> 
> The thing is that the fanart and icon are actually missing from your submission!

I've fixed the missing files.
I have also resubmitted the enum34 module as PR #1334.
It should now be compatible with Python 2 & 3 per your suggested changes.

basrieter commented on script.service.ltvpl/util.py:
```diff
@@ -0,0 +1,139 @@
+ADDON_USERDATA_FOLDER = xbmc.translatePath("special://profile/addon_data/"+ADDONID)
+BASEPATH = os.path.join(ADDON_PATH,r"resources")
+if BASEPATH not in sys.path:
+    sys.path.insert(2,BASEPATH)
```
Please don't use `sys.path`, but give your add-on a proper structure so you can use things like `from resources.lib.data...... import .....`

jmooremcc replied:
Removed the code setting sys.path

basrieter commented on script.service.ltvpl/util.py:
```diff
@@ -0,0 +1,139 @@
+    sys.path.insert(2,BASEPATH)
+ENUMPATH = os.path.join(BASEPATH,"lib2")
+if ENUMPATH not in sys.path and PYVER < 3.0:
+    sys.path.insert(0,ENUMPATH)
```
Please don't use `sys.path`, but give your add-on a proper structure so you can use things like `from resources.lib.data...... import .....`

jmooremcc replied:
Removed code modifying sys.path as well as reference to lib2 which is no longer needed for enum.

basrieter commented:
So enum34 is approved now, and is waiting to go with this one. However, we still need a bit more time here. Mainly due to the reason that:

- there are a lot of code files
- the code very hard to read 
- seems like there are many areas that are not Py2/3 compatible.

jmooremcc commented:
> So enum34 is approved now, and is waiting to go with this one. However, we still need a bit more time here. Mainly due to the reason that:
> 
> * there are a lot of code files
> * the code very hard to read
> * seems like there are many areas that are not Py2/3 compatible.

I use nextPVR for the live TV component. I've run my code under Matrix without any major problems. However, since at the time there was not a compatible nextPVR add-on available, I wasn't able to fully test it under Matrix. Complicating the issues was that some versions of Matrix had Python 2.7 installed instead of Python 3.x. If you are aware of a version of Matrix that I can download to fully test with nextPVR, I would appreciate it.

BTW, My add-on uses a client-server architecture. I've tested the server portion under Windows running Python 3.6 without any problems.

jmooremcc commented:
I've tested the backend server code under Python 3.7 and saw no compatibility problems.
All the Python code under the resources folder was included in this test.
The actual add-on code that's found outside the resources folder can only be tested under a version of Matrix that has Python 3.x that includes a compatible PVR add-on (I'm using NextPVR).

jmooremcc commented:
@basrieter @enen92 
This revision has been tested under both Leia and Matrix and runs successfully under both.
The Matrix version I tested under was the nightly build "KodiSetup-20200322-6b1585df-master-x64". The Leia version I used was "kodi-18.6-Leia-x64" which still requires the enum34 module to run.

Please let me know if you have any questions or need more information.
Thank you.

jmooremcc commented:
> Ok, so I went through the majority of your code, and I don't really know what to think? The code is very, very hard to read. And it seems that there is some kind of debug code built into it, which makes it even harder. So I added some comments, but there are many more.
> 
> So I would suggest you try to make the code more readable so we can do a proper review on it. I know that might sound like you heard that before, but the fact that this PR is sitting so long, has to do with the state it is in. So by making it more readable and understandable we can review it better. Add some comment to complex parts so we understand it.
> 
> Please don't take this personal. But we do want to be able do a proper assessment of new add-ons as it will run on many Kodi installations.

First, let me thank you for you kind words of encouragement and helpful suggestions. This is a first time Kodi add-on project for me and I'm obviously nowhere near the level of expertise you have with the Kodi system. I especially expected to get feedback on the Kodi GUI interface part of the code because I found the documentation a bit sparse and hard to understand. I actually borrowed bits and pieces of code from other developer's projects to get the GUI to work to my satisfaction.

As far as the Debugging code is concerned, I got that idea from the nextPVR add-on which prints out debugging statements if debug mode is activated in it's settings. My add-on does likewise and provides a debug mode activation setting if needed. As I'm sure you are aware, printing debugging statements is the best way to track down problems in a real time application. If a user reports a problem, the debug statements will make it easier for me to figure out what happend.

The technical assessment is one thing but I would also like to get your assessment of the add-on's functionality since how it works, from my perspective, is the most important part. I developed this functionality to assist me with my TV viewing habits and have been using it daily for over 2 years. I use the add-on to program my daily morning viewing block as well as my afternoon news and information block. For example, I have programmed my TV using this add-on to let me watch a half hour of each local station's news plus my favorite national news program. The add-on also lets me program recurring events like weekly programs I'm interested in watching.

Other Kodi users have seen this add-on working and have expressed how useful they found it to be. In one case, an individual thought it would be perfect for his grandmother, who barely can turn the TV on, since it would allow him to preset her favorite shows for her viewing pleasure.

This add-on is chock full of useful features. In fact, it checks the EPG daily to validate that day's viewing schedule so that if a program is preempted, it will not activate that program's channel change. It also has a vacation mode controlled in the settings so that the add-on will not run while you're away.

 I'm especially appreciative of your diagnosing the problem I pointed out to @ronie about my textures disappearing if I removed the library="resources" extension. I'd also like to point out that I've been getting conflicting directions about what to include in my addon.xml file and what to remove. The items you suggested that I add, were originally in my addon.xml file and I was told to remove them for whatever reason. 

Again, I want to thank you for your help, your patience and your comments and I will do my best to resolve the issues you've pointed out.

jmooremcc commented on script.service.ltvpl/resources/lib/KodiLib/kodiflags.py:
```diff
@@ -0,0 +1,11 @@
+
+try:
+    import xbmc
+    KODI_ENV = False
```
This is a client/server application and the backend was developed using Pycharm Community Edition on w Windows 10 computer. The back end is also designed to work as a stand-alone server which is why the code to determine the environment is important. 
I am also using Kodistubs and Web-Pdb but Web-Pdb is not as fast nor as efficient as the Pycharm debugger which is why I only used it for the GUI code.

jmooremcc commented on script.service.ltvpl/resources/lib/KodiLib/KodiUtilities.py:
```diff
@@ -0,0 +1,423 @@
+    :rtype Kodi
+    :return:
+    """
+    return(Kodi("http://192.168.3.107:8080/jsonrpc"))
```
I'll remove it since it was used for testing purposes.

jmooremcc commented on script.service.ltvpl/Countdown.py:
```diff
@@ -0,0 +1,482 @@
+
+    @staticmethod
+    def onErrorNotification(opstatus, errMsg):
+        xbmc.log("*****onErrorNotification {}:{}".format(opstatus, errMsg))
```
I've removed the xbmc.log call

jmooremcc commented on script.service.ltvpl/addon.xml:
```diff
@@ -0,0 +1,46 @@
+      Der Manager für TV-Wiedergabelisten stellt einen Service für Clients zur Verfügung, um Fernsehsendungen zu planen. Eine Wiedergabeliste für TV-Programme ist eine Liste von Kanalwechselbefehlen, die zu einer bestimmten Zeit stattfinden. Ein Element der Wiedergabeliste besteht aus einer Alarmzeit, einem Kanal und einer Wiederholungsoption. Der Manager verwaltet eine Datenbank mit Wiedergabeli ...
+    </description>
+    <platform>all</platform>
+    <license>GNU General Public License v3.0 only, June 2007</license>
```
I'm confused about this item. I thought I had made the correct change previously but obviously did not do so correctly. I would appreciate it if you could give me more directions or an example to go by.

jmooremcc commented on script.service.ltvpl/Countdown.py:
```diff
@@ -0,0 +1,482 @@
+        return super(Countdown, cls).__new__(cls, 'ltvpl-countdownDialog.xml', xbmcaddon.Addon(addonID).getAddonInfo('path'))
+
+    def __init__(self, addonID, client=None, listitem=None, clockstarttime=15, abortChChange=None):
+        super(Countdown, self).__init__()
```
"Why don't you use a proper resource add-on that contains your textures?"
I don't understand what you mean by this. 
Please give me an example demonstrating a "proper resource add-on".

I will work on the problem you pointed out in Countdown.py. I'm certain that the same problem exists in other GUI modules as well.

jmooremcc commented:
@basrieter Could you do me a favor and give me a list of the top 3 parts of my code that is difficult for you to read. 
Thank you.

basrieter commented:
> @basrieter Could you do me a favor and give me a list of the top 3 parts of my code that is difficult for you to read.
> Thank you.

I can give it a try: 

1. The whole structure and design is so hard to grasp: The file structure in  `resources` seems ok, but the root of the add-on completely ruins that
1. Things are over complicated. It looks like the whole `KodiLib` modules should be removed and code that you use, moved to the corresponding classes. 
1. Code is also overcomplicated: why the `KodiTransport` base class? There is just a single class that inherits from it. And what does `kodiflags.py` do? It seems to facilitate debugging? 
1. And finally the client-server setup seems over the top, but that is a choice. But it is implemented in such a chaotic way that it is almost impossible to follow. 
    1. Create a `common` module that contain common definitions of data types and components
    1. Create a `client` module that contains the client side code using the `requests` module
    1. Create a `server` module that contains all logic for the server part. You could simply use a `flask` for that.
1. You seems to prefer to inherit from classes and extend them to your needs. Some classes inherit from three base classes, such as `PL_DataSet`. This is just bad practice. Read https://en.wikipedia.org/wiki/Composition_over_inheritance to see how you could solve that. From my perspective I see most of the Kodi Python classes as _interface_ classes. I try to work with my own objects in side the code, and finally convert them to Kodi objects such as ListItems. 
1. And finally: although it seems to make little sense: try to use the PEP naming conventions. This makes things so much easier to read. Use PyCharm (you can get a free open source license for all their tools) as an IDE so it warns you about issues. 

I hope this helps you a bit. Please let me know if you have some more questions that could help you. For now, I will close this PR and hope to see a revised one in the future.
